### PR TITLE
Add Scala 3 build

### PR DESCRIPTION
**Pull request by ybasket**
Adds Scala 3.0.1 builds for all modules that can support it. Extracts `cats-tagless` instances so that they are only present on Scala 2. Introduces `sbt-projectmatrix` to cater for the fact that both `scalacache` and `akka-http` don't have Scala 3 builds (and don't seem to get some soon). Not sure that this didn't break CI release setup though, hence marking as draft.

### CODE REVIEW

File: modules/core/src/main/scala/sup/HealthResult.scala [hunk 1]
```diff
@@ -11,25 +10,21 @@ import sup.data.Tagged
...
-  }
-
   implicit def monoid[H[_]: Applicative](implicit M: Monoid[Health]): Monoid[HealthResult[H]] =
-    new Monoid[HealthResult[H]] {
-      override val empty: HealthResult[H] = HealthResult.const[H](M.empty)
-
-      override def combine(x: HealthResult[H], y: HealthResult[H]): HealthResult[H] =
-        HealthResult(Applicative.monoid[H, Health].combine(x.value, y.value))
-    }
+    Monoid.instance(
```
Comment (ybasket) [inline modules/core/src/main/scala/sup/HealthResult.scala:24]: I had some weird error on Scala complaining about a generated bridge being in conflict after type erasure, but luckily, `Monoid.instance` is a good drop-in replacement.

File: modules/core/src/main/scala/sup/data/Tagged.scala [hunk 2]
```diff
@@ -16,7 +16,9 @@ object Tagged {
     * The only place where it should be passed to is [[sup.HealthReporter.fromChecks]],
     * for determining the status of the wrapping check. In other cases, it's probably useless, as it discards the tag completely.
     */
-  implicit def catsReducibleForTagged[Tag]: Reducible[Tagged[Tag, *]] = Reducibles.by(λ[Tagged[Tag, *] ~> Id](_.health))
+  implicit def catsReducibleForTagged[Tag]: Reducible[Tagged[Tag, *]] = Reducibles.by(new (Tagged[Tag, *] ~> Id) {
```
Comment (ybasket) [inline modules/core/src/main/scala/sup/data/Tagged.scala:19]: `FunctionK` lambda syntax is not supported on Scala 3, hence the more verbose rewrite.

File: modules/redis/src/main/scala/sup/modules/redis.scala [hunk 3]
```diff
@@ -2,6 +2,7 @@ package sup.modules
 
 import cats.implicits._
 import cats.{ApplicativeError, Id}
+import cats.catsInstancesForId
```
Comment (ybasket) [inline modules/redis/src/main/scala/sup/modules/redis.scala:5]: Due to changes in implicit resolution on Scala 3, this becomes necessary.

File: project/plugins.sbt [hunk 4]
```diff
@@ -4,3 +4,4 @@ addSbtPlugin("com.eed3si9n" % "sbt-buildinfo" % "0.10.0")
 addSbtPlugin("com.typesafe" % "sbt-mima-plugin" % "0.9.2")
 addSbtPlugin("io.github.davidgregory084" % "sbt-tpolecat" % "0.1.20")
 addSbtPlugin("com.codecommit" % "sbt-github-actions" % "0.12.0")
+addSbtPlugin("com.eed3si9n" % "sbt-projectmatrix" % "0.8.0")
```
Comment (kubukoz) [inline project/plugins.sbt:7]: Not sure how I feel about this... Do we actually gain anything with this plugin?

File: modules/core/src/main/scala-2/sup/HealthResultTaglessInstances.scala [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package sup
+
+import cats.tagless.FunctorK
+import cats.~>
+
+trait HealthResultTaglessInstances {
```
Comment (kubukoz) [inline modules/core/src/main/scala-2/sup/HealthResultTaglessInstances.scala:6]: Strange, I thought cats-tagless-core would be available for Scala 3 (it doesn't have macros AFAIK). Maybe it's worth making a ticket to see if/when that could happen :)

**Comment by kubukoz** [comment]
Wrong button 😂

**Comment by kubukoz** [comment]
Thanks! Hopefully the CI will work, although I'm worried disabling certain projects under 3.x might do some evil there.

File: project/plugins.sbt (hunk 4 above)
Comment (kubukoz) [reply project/plugins.sbt:7]: I see - this probably makes it easier to disable some subprojects. Hopefully CI likes it :)

File: modules/core/src/main/scala-2/sup/HealthResultTaglessInstances.scala (hunk 5 above)
Comment (ybasket) [reply modules/core/src/main/scala-2/sup/HealthResultTaglessInstances.scala:6]: There's a somewhat stale issue suggesting that it requires some CI updates first (and macros are blocked by Scala itself): https://github.com/typelevel/cats-tagless/issues/170

**Comment by ybasket** [comment]
> Thanks! Hopefully the CI will work, although I'm worried disabling certain projects under 3.x might do some evil there.

It didn't, but just because I forgot to fix the tests 🙈 Could trigger for the new version?

And yes, `sbt-projectmatrix` could totally cause problems, but I'm not aware of another (good) possibility to handle this situation, just setting `crossScalaVersions` on the special modules doesn't work. Let's see what CI says…

**Comment by kubukoz** [comment]
Could you make some small boring PR to main so I don't have to approve workflows every time? 😂😂

**Comment by ybasket** [comment]
> Could you make some small boring PR to main so I don't have to approve workflows every time? 😂😂

Be careful what you ask for, you might get it 😉: https://github.com/kubukoz/sup/pull/365

But CI on this one actually went green with the test fix. Might be worth to do a snapshot release or alike to test it though…hard to say upfront whether `sbt-projectmatrix` and `sbt-ci-release` or so interact weirdly.

**Comment by kubukoz** [comment]
I'm good with merging to main, if it breaks we can work on it later :) I think we have snapshots on...

**Comment by kubukoz** [comment]
Thanks for working on this!

**Comment by kubukoz** [comment]
There's some issue while building docs. I've seen this before in 3.0.1, but not in 3.0.0, can you downgrade in another PR?

**Comment by ybasket** [comment]
> There's some issue while building docs. I've seen this before in 3.0.1, but not in 3.0.0, can you downgrade in another PR?

Sure, here you go: https://github.com/kubukoz/sup/pull/366

**Comment by kubukoz** [comment]
Another one, looks like after this we're building all Scala versions in each job (so everything gets built 3 times):

image

Probably some setup is needed to tell sbt-github-actions how to build correctly in the presence of sbt-project-matrix.

**Pull request merged**